src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -473,6 +482,18 @@
   (when emit-cmd-events?
     (async/>!! cmd-event-ch (queue/make-cmd-event cmdref kind))))
 
+(defn check-schema-version [desired-schema-version context db service]
+  ;; should we guard against concurrent gc runs here too?
+  (when-not (= #{:stopping} @(:stop-status context))
+    (let [schema-version (-> (jdbc/with-transacted-connection db
+                               (jdbc/query "select max(version) from schema_migrations"))
+                             first
+                             :max)]
+      (when (> schema-version desired-schema-version)
+        ((:shutdown-on-error context) (service-id service)
+         #(throw (ex-info (str (trs "Schema has been upgraded by another PuppetDB"))
+                          {:kind ::schema-mismatch})))))))
```

Comment:
Need to update the error checking in `start-puppetdb` once #3204  is merged to account for this error type.